text/0066-open-record.md

```diff
@@ -0,0 +1,164 @@
+# The default attribute type for open records
+
+## Related issues and PRs
+
+- Reference Issues: [#305](https://github.com/cedar-policy/cedar/issues/305)
+- Supercedes [RFC 27](https://github.com/cedar-policy/rfcs/pull/27)
+- Implementation PR(s): (leave this empty)
+
+## Timeline
+
+- Start Date: 2024-05-22
+- Date Entered FCP:
+- Date Accepted:
+- Date Landed:
+
+## Summary
+
+The Cedar validator requires a user to enumerate all possible record and entity attributes, both names and types, in the schema. This is limiting when you'd like to treat a record like a map, i.e., there are attributes whose names are not known in advance, but whose value types are.
+
+This RFC proposes to extend the validator to support _open records_. An open record type extends a normal record type with a _default attribute type_. Policies can check for the presence of any attribute in an open record using `has`, and if present they can access it assuming its value has the default attribute type.
+
+## Motivating example: Tags as Attribute-based Maps
+
+As a key motivating example, you can use open records to implement _tags_, which are basically a map from `String` keys to values.
+
+For example, suppose that `principal.tags` has type `{ } default String`, meaning it is a record with default attribute type `String`. Then you can write `principal.tags has "priority" && principal.tags["priority"] == "green"`. Just writing `principal.tags["priority"] == "green"` without a preceding `has` check would result in a validation error.
+
+Compare this approach to another way of implementing tags, as sets of key-value pairs. For example, `principal.tags` could have type `Set<{ key:String, value:String }>` and a tag check would involve an expression like `principal.tags.contains({ key: "priority", value: "green" })`. Attribute-based tags are useful when you want to compare tags on different entities, as in 
+`resource.tags["project"] == principal.tags["project"]` 
+or 
+`resource.tags["projects"].contains(principal.tags["project"])` (when tags have type `{ } default Set<String>`, so tags' values are sets of strings). Doing this is not possible when encoding tags as sets.
+
+Compared to the status quo, open records allow you to implement attribute-based tags without changing the schema each time you add a tag. See the [Workarounds](#workarounds) section below for discussion.
+
+## Detailed design
+
+Supporting open records will requre extensions to the Cedar schema format and the validator. Neither the JSON format for entities nor the JSON or natural syntax for policies needs to change to support open record objects.
```

Comment:
> Every component of Cedar will be affected, including partial evaluation and analysis.

I agree that analysis is affected inasmuch as it relies on validation.

I don't understand why partial evaluation is affected, because in my understanding partial evaluation is about evaluation, defined independently of validation. And there is nothing about evaluation that is affected by this RFC. What am I missing?

> (core datastructures need to change).

Can you elaborate on why this is necessary? Again, on the face of it we are only changing the validator, not the evaluator.

Reply:
Partial evaluation will have to deal with open records too.  

Basically, implementing this requires changing the definition of `Cedar.Spec.Value`, which is a core definition in the model that everything depends on.

Plus, PE models will also have to change their representation of partial values to support open records.

Analysis does not depend on validation, but it will have to fundamentally change to support open records.

Reply:
Ah, I see that this is _not_ suggesting adding _default_ at the `Value` level, just the type level.  And we can get away with this because of dynamic semantics for concrete and partial evaluation.

Unfortunately, the difficulties with analysis still remain because it's type-directed.  (It will have to change significantly.)